### PR TITLE
Use environment variable and client option for API key

### PR DESCRIPTION
**Pull request by dgellow**
Ref: https://github.com/stainless-api/stainless/pull/89

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,8 +16,10 @@ yarn add lithic
 ```js
 import Lithic from 'lithic';
 
-const lithic = new Lithic(process.env.LITHIC_API_KEY, {
-  environment: 'sandbox', // or 'production'
+const lithic = new Lithic({
+  apiKey:
+    'my api key' /* defaults to process.env["LITHIC_API_KEY"]. Set to null to send unauthenticated requests */,
```
Comment (dgellow) [inline README.md:21]: I'm not happy that I'm using `/* ... */` here but with the current implementation the extra `,` at the end of the line is automatically added, which is incorrect when using `// ...`.

And it's a bit of work to fix that specific point.

Comment (rattrayalex) [reply README.md:21]: Let's omit this for now:

> Set to null to send unauthenticated requests

It's a rare edge case and shouldn't be the first thing people see. 

Maybe instead we can add it to a docstring in the option [here](https://github.com/stainless-api/lithic-node/pull/9/files#diff-dcdc3e0b3362edb8fec2a51d3fa51f8fb8af8f70247e06d9887fa934834c9122R16)

Comment (dgellow) [reply README.md:21]: > Maybe instead we can add it to a docstring in the option [here](https://github.com/stainless-api/lithic-node/pull/9/files#diff-dcdc3e0b3362edb8fec2a51d3fa51f8fb8af8f70247e06d9887fa934834c9122R16)

That's a good idea

Comment (dgellow) [reply README.md:21]: > Ugh… no I think we need to invest the work and fix this in our codegen systems

For sure, but would you say that blocks the PR https://github.com/stainless-api/stainless/pull/89? Just to know if you would prefer I spend some time on it now or if we merge it as it is and open a ticket to fix this issue at a later time.

File: index.ts [hunk 2]
```diff
@@ -38,10 +43,10 @@ export class Lithic extends Core.APIClient {
...
 
   protected override defaultHeaders(): Core.Headers {
-    const Authorization = `api-key ${this.apiKey}`;
+    const authHeader = { Authorization: `api-key ${this.apiKey}` };
     return {
       ...super.defaultHeaders(),
-      Authorization,
+      ...(this.apiKey ? authHeader : {}),
     };
   }
```
Comment (dgellow) [inline index.ts:51]: Here

File: README.md (hunk 1 above)
Comment (rattrayalex) [reply README.md:21]: I don't think it's okay to ship with the block comments. We can sidestep the codegen system for this code example now if we like, that might be simpler.

Comment (dgellow) [reply README.md:21]: Cool, thanks for the quick response, I will look into it then 👍.

Comment (rattrayalex) [reply README.md:21]: No worries! Sorry for the scattered convo.

Shouldn't be too bad to peel back a layer of abstraction or two here for now, though ultimately we really do want to be able to add comments to things like this somewhat arbitrarily…

File: index.ts [hunk 3]
```diff
@@ -6,20 +6,28 @@ import type { Agent } from 'http';
...
+  constructor({
+    apiKey = process.env['LITHIC_API_KEY'],
+    environment = 'production',
+    baseURL,
+    timeout,
+    httpAgent,
+  }: {
+    /**
+     * Defaults to to process.env["LITHIC_API_KEY"]. Set it to null if you want to send unauthenticated requests.
+     */
```
Comment (dgellow) [inline index.ts:18]: @rattrayalex Relevant change

File: README.md [hunk 4]
```diff
@@ -97,13 +99,13 @@ You can use the `maxRetries` option to configure or disable this:
...
 // Configure the default for all requests:
-const lithic = new Lithic(process.env.LITHIC_API_KEY, {
+const lithic = new Lithic({
   maxRetries: 0, // default is 2
 });
 
 // Or, configure per-request:
 lithic.cards.list({ page_size: 10 }, {
-  maxRetries: 5
+  maxRetries: 5,
```
Comment (dgellow) [inline README.md:108]: @rattrayalex  Small change here and for other expanded objects. We were not too consistent before, now we always have a comma for expanded properties.

Comment (rattrayalex) [reply README.md:108]: Nice!

File: index.ts (hunk 3 above)
Comment (rattrayalex) [reply index.ts:18]: Nice!

File: index.ts [hunk 5]
```diff
@@ -6,20 +6,28 @@ import type { Agent } from 'http';
...
+    apiKey?: string | null;
```
Comment (rattrayalex) [inline index.ts:19]: I'm realizing we should allow `apiKey` and `environment` in `RequestOptions` too, but that can be a follow-on

**Review by rattrayalex — APPROVED** [review]

File: tests/index.test.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
+// File generated from our OpenAPI spec by Stainless.
+
+import Lithic from '../index';
+
+describe('instantiate client', () => {
```
Comment (dgellow) [inline tests/index.test.ts:5]: @rattrayalex I went further and added a bunch of tests to cover the various way the client can be instantiated.

File: index.ts [hunk 7]
```diff
@@ -5,26 +5,70 @@ import type { Agent } from 'http';
 
 const environments = { production: 'https://api.lithic.com/v1', sandbox: 'https://sandbox.lithic.com/v1' };
 
+type Config = {
```
Comment (dgellow) [inline index.ts:8]: @rattrayalex Relevant changes

File: tests/index.test.ts (hunk 6 above)
Comment (rattrayalex) [reply tests/index.test.ts:5]: Nice, I really like these!

Comment (dgellow) [reply tests/index.test.ts:5]: Something that came to my mind this morning: I haven't added a test for `environment` and `baseUrl`. @rattrayalex could you check I haven't broken anything here? With a test or via manual testing.

**Pull request merged**